fusestore/store/src/executor/kv_handlers.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020-2021 The Datafuse Authors.
+//
+// SPDX-License-Identifier: Apache-2.0.
+//
+
+use common_exception::ErrorCode;
+use common_flights::GetKVAction;
+use common_flights::UpsertKVAction;
+use common_store_api::GetKVActionResult;
+use common_store_api::UpsertKVActionResult;
+
+use crate::executor::action_handler::RequestHandler;
+use crate::executor::ActionHandler;
+use crate::meta_service::ClientRequest;
+use crate::meta_service::ClientResponse;
+use crate::meta_service::Cmd;
+
+#[async_trait::async_trait]
+impl RequestHandler<UpsertKVAction> for ActionHandler {
+    async fn handle(&self, act: UpsertKVAction) -> common_exception::Result<UpsertKVActionResult> {
+        let cr = ClientRequest {
+            txid: None,
+            cmd: Cmd::UpsertKV {
+                key: act.key,
+                seq: act.seq,
+                value: act.value,
+            },
+        };
+        // TODO(xp): raftmeta should use ErrorCode instead of anyhow::Error
+        let rst = self
+            .meta_node
+            .write(cr)
+            .await
+            .map_err(|e| ErrorCode::MetaNodeInternalError(e.to_string()))?;
+
+        match rst {
+            ClientResponse::KV { prev, result } => Ok(UpsertKVActionResult { prev, result }),
+            _ => Err(ErrorCode::MetaNodeInternalError("not a KV result")),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl RequestHandler<GetKVAction> for ActionHandler {
+    async fn handle(&self, act: GetKVAction) -> common_exception::Result<GetKVActionResult> {
+        let result = self.meta_node.get_kv(&act.key).await;
+        Ok(GetKVActionResult { result })
+    }
+}
```

Comment:
An abstraction may be better to be done on the `action`, e.g., `UpserAction` instead of on the `ActionHandler`
Because `ActionHandler` is no more than a request router,  and actual work is done by `MetaNode`, and business logic intuitively, is bound to a specific `Action`.

```rust
trait Handle {
  type Output;
  fn handle(self, meta_node: &MetaNode) -> Result<Self::Output>;
}

impl Handle for UpsertAction {
  type Output = UpserActionResult;
  fn handle(...)
}
```

After all, I'd prefer concrete codes than abstraction.

Reply:
> Because ActionHandler is no more than a request router
yes, but named as `RequestHandler`:
~~~
pub trait RequestHandler<T>: Sync + Send
where T: RequestFor
{
    async fn handle(&self, req: T) -> common_exception::Result<T::Reply>;
}
~~~

Instead of a simple `type Outut`, another trait `RequestFor` is introduced here, which binds Request/Reply types together.

> ActionHandler is no more than a request router, and actual work is done by MetaNode, and business logic intuitively, is bound to a specific Action.

yes, so we have 

` impl RequestHandler<UpsertKVAction> for ActionHandler ... `

`impl RequestHandler<GetKVAction> for ActionHandler .... `

with the help of trait `RequestFor` 

the `handle` method can only return the type which a `RequestFor<UpsertKVAction>` requests for, i.e. `UpsertKVActionResult`

~~~

impl RequestHandler<UpsertKVAction> for ActionHandler {
    async fn handle(&self, act: UpsertKVAction) -> common_exception::Result<UpsertKVActionResult> {
    
~~~

The handler returns Result<UpsertKVActionResult> instead of :

~~~
                Ok(StoreDoActionResult::UpsertKV(UpsertKVActionResult {
                    prev,
                    result,
                }))    
~~~

and, in fact, before refactoring, if we return A DoActionResult of another variant, like                   

~~~
Ok(StoreDoActionResult::UpsertKV(GetKVActionResult { ... }))    
~~~

The compiler will accept it happily (and leads to runtime error)

After refactoring, this type of error will be detected by the compiler.

Reply:
I am open to naming rules, any suggestions are welcome

Reply:
Do you still allow polymorphism in your style? I remember java web programming has similar issues.
Does your approach also apply to java, or it is something unique to Rust?

Reply:
After the refactoring, no.

A `RequestHandler` should return the type which is requested for (trait `RequestFor` encodes the req/reply relation)


As for java, if I recall correctly, can also express such constraints